lib/helpers/versions.js

```diff
@@ -2,15 +2,84 @@ const logger = require('winston');
 const semver = require('semver');
 const stable = require('semver-stable');
 const _ = require('lodash');
+const semverUtils = require('semver-utils');
 
 module.exports = {
+  determineRangeUpgrades,
   determineUpgrades,
   isRange,
   isValidVersion,
   isFuture,
   isPastLatest,
+  computeRangedVersion,
 };
 
+function determineRangeUpgrades(dep, currentVersion, config) {
+  if (!isValidVersion(currentVersion)) {
+    logger.verbose(`${dep.name} currentVersion is invalid`);
+    return [];
+  }
+  const versions = dep.versions;
+  if (!versions || Object.keys(versions).length === 0) {
+    logger.verbose(`${dep.name} - no versions`);
+    return [];
+  }
+  const versionList = Object.keys(versions);
+  const workingVersion = semver.maxSatisfying(versionList, currentVersion);
+
+  return _.chain(versionList)
+
+    // Filter out older versions as we can't upgrade to those
+    .filter(version => semver.gt(version, workingVersion))
+
+    // Ignore unstable versions, unless the current version is unstable
+    .reject(version => config.ignoreUnstable &&
+            stable.is(workingVersion) && !stable.is(version))
+
+    // Ignore future versions, unless the current version is marked as future
+    .reject(version => config.ignoreFuture &&
+            !isFuture(versions[workingVersion]) && isFuture(versions[version]))
+
+    // Ignore versions newer than "latest", unless current version is newer than the "latest"
+    .reject(version => config.respectLatest &&
+            isPastLatest(dep, version) && !isPastLatest(dep, workingVersion))
+
+    // Process all remaining versions
+    .reduce((upgrades, newVersion) => {
```

Comment:
@destroyerofbuilds I think I know how to solve this using a semver range "hack". We could do it like this:

- Find the maximum *version* for each major version (e.g. existing might be `~1.5.0` and we find `1.7.6`). This could be the same logic already used before this PR
- Convert this maximum version to a minor semver range by dropping the last digit, e.g. `1.7`. In semver speak, `1.7` is equivalent to `1.7.x` which for our purposes is equivalent to a tilde range
- Use `semver`'s `minSatisfying` by passing it the full array of versions and `1.7`. This usually would then return `1.7.0` but could also return `1.7.1` for example, if `1.7.0` is missing.

In theory we would be iterating over the versions array twice to do this, but I think the performance impact would be so minimal that there's no point worrying about it.